### ISSUE TITLE
0.22: "stg mail" generated cover letter is not correct

### ISSUE THREAD

**Issue by chucklever** [ctmarinas/stgit#55]
[cel@klimt linux]$ stg version
Stacked GIT 0.22
git version 2.26.0.rc1.11.g30e9940356dc
Python version 2.7.5 (default, Apr 11 2018, 17:41:36) 
[GCC 4.8.5 20150623 (Red Hat 4.8.5-28.0.1)]
[cel@klimt linux]$

After the cover letter is constructed, shortlog shows the short description for only the last patch in the mailed series, and the cover letter's diffstat is performed on the entire branch (or perhaps all applied patches?) instead of only the requested patches.

I'm guessing that 5b99accc4e92 ("Migrate `stg mail` to new lib infrastructure") is the cause of both problems.

The shortlog issue appears to be that "git log" does not understand multiple commits on its command line. It outputs a log entry only for the last "commit ^commit" pair. Changing the __shortlog function to use "git show" instead of "git log" fixes the problem, and then the for loop needs to be changed to process the patch list in reverse order.

For the diffstat issue, I'm not smart enough to figure out a solution, but this new bit of code looks incorrect:

```
        'diffstat': repository.diff_tree(
            stack.base.data.tree,
            stack.top.data.tree,
            diff_opts=options.diff_flags,
            stat=True,
```

The diffed trees need to be derived from the patches that are being mailed, not from the base and top of the current stack.

**Comment by chucklever** [comment]
Another approach for __shortlog would be to stick with "git log" and simply pass it the range of commit IDs. This approach also addresses the diffstat issue:

```
def __diffstat(stack, patches, diff_opts):
    rev1 = stack.patches.get(patches[0])
    rev2 = stack.patches.get(patches[-1])
    return stack.repository.diff_tree(
            rev1.commit.data.tree,
            rev2.commit.data.tree,
            diff_opts,
            stat=True,
    )
```

**Comment by chucklever** [comment]
My Python is immensely rusty, but here's a patch that addresses the issues.

https://github.com/chucklever/stgit/commit/d120b78ab791d467a1c14ae894142de5c6971c70

**Comment by chucklever** [comment]
My approach works fine for ranges, but providing "stg mail" with a list of individual patches will require __shortlog to continue to use a loop that processes each patch in the list. I will update my patch to address this.

Unfortunately, a list-based approach works for the diffstat: There's no way to make that output meaningful for a sparse list of patches. For the moment, I'm sticking with a range-based approach, which was what was in place before https://github.com/ctmarinas/stgit/commit/5b99accc4e927b7fce4b964468e9fb4917f3b7ba .

**Comment by chucklever** [comment]
I forgot to update here last week. The fix has been split into two commits:

shortlog:
https://github.com/chucklever/stgit/commit/5ad3d67035132122986c9b5ae3ce0284c6f5f120

diffstat:
https://github.com/chucklever/stgit/commit/aad399762a4dc359d7e4846180dfd3e84de5b48e

**Comment by jpgrayson** [comment]
Thanks @chucklever for this issue report and for these patches. I have not yet had time to dig into the details, but I appreciate this help and I will look into this deeper in my next StGit cycles.

I will admit that `stg mail` is a bit of a blind spot for me. The test suite does not cover `stg mail` particularly well and my personal workflows have not needed `stg mail`. A longer-term goal I have in mind for `stg mail` is to see if we can replace its guts with `git send-email`. My general belief is that `stg mail` could shed a lot of code and become more correct by leveraging `git send-email`.

So it's nice to know that there is at least one `stg mail` user out there!

**Comment by chucklever** [comment]
The one open issue for me here is I would like to know how to generate a proper diffstat for a sparse list of patches! Perhaps git send-email would help with that.

**Comment by chucklever** [comment]
Fwiw, the diffstat situation has not gone unnoticed within git itself:

https://lwn.net/Articles/802797/

**Comment by jpgrayson** [comment]
I have pushed the two repairs you made, @chucklever. Thank you!

I updated a unit test to check for the shortlog problem.

And I made one change to *not* pass the user-provided diff options (from `--diff-opt`) to the diff stat in the shortlog. I think those options should only be applied to the diffs in the patch messages themselves. Instead, we now pass `--stat-width=72` to for the shortlog's diffstat. This is similar to what `git format-patch` does.

Anyway, I think this issue can be put to rest. Thank you again for finding and fixing this problem!
